### PR TITLE
Probe: Fix early-exit of debuggee

### PR DESCRIPTION
**Pull request by krf**
Start to add in-process integration tests

### CODE REVIEW

**Comment by krf** [comment]
Supersedes #320

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -208,6 +208,18 @@ endif()
 
 endif()
 
+### Integration test
+
+add_executable(integrationtest
+  integrationtest.cpp
+  ../probe/probecreator.cpp
```
Comment (milianw) [inline tests/CMakeLists.txt]: why not link against gammaray_probe?

File: core/probe.cpp [hunk 2]
```diff
@@ -396,6 +398,11 @@ void Probe::delayedInit()
         showInProcessUi();
 }
 
+void Probe::shutdown()
+{
+    delete this;
```
Comment (milianw) [inline core/probe.cpp]: yikes ;-) usually, this is a very bad idea, but here there's really no way around this, is there?

File: core/probe.cpp [hunk 3]
```diff
@@ -294,7 +291,7 @@ Probe *GammaRay::Probe::instance()
 
 bool Probe::isInitialized()
 {
-    return instance() && qApp;
+    return instance();
```
Comment (milianw) [inline core/probe.cpp]: why is this change necessary?

File: core/probe.h [hunk 4]
```diff
@@ -56,6 +56,14 @@ class BenchSuite;
 class Server;
 class ToolManager;
 
+/**
```
Comment (milianw) [inline core/probe.h]: these comments could in principle be separate commits

File: core/probe.cpp (hunk 2 above)
Comment (vkrause) [inline core/probe.cpp]: The alternative I see for this is connect(qApp, &QCoreApp::aboutToQuit, [](){ delete Probe::instance(); }), but that's Qt5-only.

File: tests/CMakeLists.txt (hunk 1 above)
Comment (krf) [inline tests/CMakeLists.txt]: Copied from other tests in that file

File: core/probe.cpp (hunk 3 above)
Comment (krf) [inline core/probe.cpp]: `instance()` now implies that a `qApp` exists, thus this check is no longer necessary. I updated the documentation of the Probe to make sure it's clear there's a strong coupling between Probe & QCoreApplication lifetime.

File: core/probe.h (hunk 4 above)
Comment (krf) [inline core/probe.h]: There's no tight coupling between Probe lifetime & QCoreApplication lifetime before (because of the deleteLater), thus this comment does not make sense without the code changes.

File: core/probe.cpp (hunk 2 above)
Comment (milianw) [inline core/probe.cpp]: Reading http://stackoverflow.com/questions/550189/is-it-safe-to-delete-this I think it's fine to have it this way though

**Comment by milianw** [comment]
+1 from my side, thanks a lot for the clarifications kevin

**Comment by krf** [comment]
Pushed.
